src/services/upload-manager.ts

```diff
@@ -19,13 +20,32 @@ export interface Uploader {
 
 export class UploadManager extends EventEmitter {
   doUpload = async (places: Place[], chtApi: ChtApi, ignoreWarnings: boolean = false) => {
-    const placesNeedingUpload = places.filter(p => !p.isCreated && !p.hasValidationErrors && (ignoreWarnings || !p.warnings.length));
+    const validPlaces = places.filter(p => {
+      return !p.hasValidationErrors && (ignoreWarnings || !p.warnings.length);
+    });
+    const placesNeedingUpload = validPlaces.filter(p => !p.isCreated);
     this.eventedPlaceStateChange(placesNeedingUpload, PlaceUploadState.SCHEDULED);
 
-    const independants = placesNeedingUpload.filter(p => !p.isDependant);
-    const dependants = placesNeedingUpload.filter(p => p.isDependant);
+    const independants = placesNeedingUpload.filter(p => !p.isDependant && !p.hasSharedUser);
+    const dependants = placesNeedingUpload.filter(p => p.isDependant && !p.hasSharedUser);
+
     await this.uploadPlacesInBatches(independants, chtApi);
     await this.uploadPlacesInBatches(dependants, chtApi);
+    await this.uploadGrouped(validPlaces.filter(p => p.hasSharedUser), chtApi);
+  };
+
+  uploadGrouped =  async (places: Place[], api: ChtApi) => {
+    const grouped = _.groupBy(places, place => place.contact.id);
```

Comment:
should this use `getPlacesByContacts` interface?

Reply:
I don't think so, `doUpload` already gets the whole place list so all we have to do is filter and group